sockeye/rnn.py

```diff
@@ -55,3 +60,68 @@ def get_stacked_rnn(cell_type: str,
             # TODO(fhieber): add pervasive dropout?
             rnn.add(mx.rnn.DropoutCell(dropout, prefix=cell_prefix + "_dropout"))
     return rnn
+
+
+class LnLSTMCell(mx.rnn.LSTMCell):
+    """
+    Long-Short Term Memory (LSTM) network cell with layer normalization.
+    Based on Jimmy Lei Ba et al: Layer Normalization (https://arxiv.org/pdf/1607.06450.pdf)
+
+    :param num_hidden: number of RNN hidden units. Number of units in output symbol.
+    :param prefix: prefix for name of layers (and name of weight if params is None).
+    :param params: RNNParams or None. Container for weight sharing between cells. Created if None.
+    :param forget_bias: bias added to forget gate, default 1.0. Jozefowicz et al. 2015 recommends setting this to 1.0.
+    :param norm_scale: scale/gain for layer normalization.
+    :param norm_shift: shift/bias after layer normalization.
+    """
+
+    def __init__(self,
+                 num_hidden: int,
+                 prefix: str = 'lstm_',
+                 params: Optional[mx.rnn.RNNParams] = None,
+                 forget_bias: float = 1.0,
+                 norm_scale: float = 1.0,
+                 norm_shift: float = 0.0) -> None:
+        super(LnLSTMCell, self).__init__(num_hidden, prefix, params, forget_bias)
+
+        self._norm_layers = list()  # type: List[LayerNormalization]
+        for name in ['i', 'f', 'c', 'o', 's']:
+            scale = self.params.get('%s_shift' % name, shape=(num_hidden,),
+                                    init=mx.init.Constant(value=norm_shift))
+            shift = self.params.get('%s_scale' % name, shape=(num_hidden,),
+                                    init=mx.init.Constant(value=norm_scale))
+            self._norm_layers.append(
+                LayerNormalization(num_hidden, prefix="%s%s" % (self._prefix, name), scale=scale, shift=shift))
+
+    def __call__(self, inputs, states):
+        self._counter += 1
+        name = '%st%d_' % (self._prefix, self._counter)
+        i2h = mx.sym.FullyConnected(data=inputs, weight=self._iW, bias=self._iB,
+                                    num_hidden=self._num_hidden * 4,
+                                    name='%si2h' % name)
+        h2h = mx.sym.FullyConnected(data=states[0], weight=self._hW, bias=self._hB,
+                                    num_hidden=self._num_hidden * 4,
+                                    name='%sh2h' % name)
+        gates = i2h + h2h
```

Comment:
Interesting! I hadn't seen this supplementary material but had looked at some other implementations. We could probably do it even more efficient by having only 3 LayerNormalization instances: 2 before the slice in line 106, and 1 for next_c in line 125 (not touching next_c itself which is returned.